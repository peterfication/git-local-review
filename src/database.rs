use refinery::config::{Config, ConfigDbType};
use sqlx::{SqlitePool, sqlite::SqliteConnectOptions};

use crate::migrations;

pub struct Database {
    pool: SqlitePool,
}

impl Database {
    pub async fn new() -> color_eyre::Result<Self> {
        std::fs::create_dir_all("tmp")?;

        let options = SqliteConnectOptions::new()
            .filename("tmp/reviews.db")
            .create_if_missing(true);

        let pool = SqlitePool::connect_with(options).await?;

        // Run migrations using refinery
        let mut config = Config::new(ConfigDbType::Sqlite).set_db_path("tmp/reviews.db");
        migrations::runner()
            .run(&mut config)
            .map_err(|e| color_eyre::eyre::eyre!("Migration failed: {}", e))?;

        log::info!("Database initialized at tmp/reviews.db with migrations");

        Ok(Self { pool })
    }

    #[cfg(test)]
    pub fn from_pool(pool: SqlitePool) -> Self {
        Self { pool }
    }

    pub fn pool(&self) -> &SqlitePool {
        &self.pool
    }

    pub async fn close(self) -> color_eyre::Result<()> {
        self.pool.close().await;
        Ok(())
    }
}
