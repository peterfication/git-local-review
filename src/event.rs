use std::sync::Arc;
use std::time::Duration;

use color_eyre::eyre::OptionExt;
use futures::{FutureExt, StreamExt};
use ratatui::crossterm::event::{Event as CrosstermEvent, KeyEvent};
use tokio::sync::mpsc;

use crate::{
    models::Review,
    services::state_service::{AppState, StateEvent},
    services::{GitBranchesLoadingState, ReviewCreateData, ReviewsLoadingState},
    views::KeyBinding,
};

/// Type alias for review identifiers to make event signatures more descriptive.
pub type ReviewId = str;

/// The frequency at which tick events are emitted.
const TICK_FPS: f64 = 30.0;

/// Representation of all possible events.
#[derive(Clone, Debug)]
pub enum Event {
    /// An event that is emitted on a regular schedule.
    ///
    /// Use this event to run any code which has to run outside of being a direct response to a user
    /// event. e.g. polling exernal systems, updating animations, or rendering the UI based on a
    /// fixed frame rate.
    Tick,
    /// Crossterm events.
    ///
    /// These events are emitted by the terminal.
    Crossterm(CrosstermEvent),
    /// Application events.
    ///
    /// Use this event to emit custom events that are specific to your application.
    App(AppEvent),
}

/// Application events.
///
/// You can extend this enum with your own custom events.
#[derive(Clone, Debug)]
pub enum AppEvent {
    /// Initialization event.
    Init,
    /// Quit the application.
    Quit,
    /// Close the current view.
    ViewClose,

    /// Trigger loading of reviews.
    ReviewsLoad,
    /// Load the reviews from the database
    ReviewsLoading,
    /// Propagates the current loading state of reviews.
    ReviewsLoadingState(ReviewsLoadingState),
    /// Load a single review by ID.
    ReviewLoad(Arc<ReviewId>),
    /// Single review loaded successfully.
    ReviewLoaded(Arc<Review>),
    /// Review was not found.
    ReviewNotFound(Arc<ReviewId>),
    /// Error occurred while loading a review.
    ReviewLoadError(Arc<ReviewId>),
    /// Inform that a review has been created.
    ReviewCreated(Review),
    /// Error occurred while creating a review.
    ReviewCreatedError(Arc<str>),
    /// Delete the selected review.
    ReviewDelete(Arc<ReviewId>),
    /// Inform that a review has been deleted.
    ReviewDeleted,
    /// Error occurred while deleting a review.
    ReviewDeletedError(Arc<ReviewId>),

    /// Open help modal with keybindings.
    HelpOpen(Arc<[KeyBinding]>),
    /// Key selected from help modal.
    HelpKeySelected(Arc<KeyEvent>),

    /// Open the review creation view.
    ReviewCreateOpen,
    /// Submit the review creation form.
    ReviewCreateSubmit(Arc<ReviewCreateData>),

    /// Open delete confirmation dialog for selected review.
    ReviewDeleteConfirm(Arc<ReviewId>),

    /// Open review details view.
    ReviewDetailsOpen(Arc<ReviewId>),

    /// Trigger loading of Git branches.
    GitBranchesLoad,
    /// Load the Git branches
    GitBranchesLoading,
    /// Propagates the current loading state of Git branches.
    GitBranchesLoadingState(GitBranchesLoadingState),

    /// State service events for centralized state management.
    StateEvent(StateEvent),
    /// Broadcast updated application state to all views.
    StateUpdate(Arc<AppState>),
}

/// Terminal event handler.
#[derive(Debug)]
pub struct EventHandler {
    /// Event sender channel.
    sender: mpsc::UnboundedSender<Arc<Event>>,
    /// Event receiver channel.
    receiver: mpsc::UnboundedReceiver<Arc<Event>>,
}

impl Default for EventHandler {
    fn default() -> Self {
        panic!("Use EventHandler::new() instead of Default");
    }
}

impl EventHandler {
    /// Constructs a new instance of [`EventHandler`] and spawns a new thread to handle events.
    pub fn new() -> Self {
        let (sender, receiver) = mpsc::unbounded_channel();
        let actor = EventTask::new(sender.clone());
        tokio::spawn(async { actor.run().await });
        Self { sender, receiver }
    }

    /// Constructs a new instance of [`EventHandler`] for testing without spawning the event task.
    /// This allows tests to control event flow manually.
    #[cfg(test)]
    pub fn new_for_test() -> Self {
        let (sender, receiver) = mpsc::unbounded_channel();
        Self { sender, receiver }
    }

    /// Receives an event from the sender.
    ///
    /// This function blocks until an event is received.
    ///
    /// # Errors
    ///
    /// This function returns an error if the sender channel is disconnected. This can happen if an
    /// error occurs in the event thread. In practice, this should not happen unless there is a
    /// problem with the underlying terminal.
    pub async fn next(&mut self) -> color_eyre::Result<Arc<Event>> {
        self.receiver
            .recv()
            .await
            .ok_or_eyre("Failed to receive event")
    }

    /// Queue an app event to be sent to the event receiver.
    ///
    /// This is useful for sending events to the event handler which will be processed by the next
    /// iteration of the application's event loop.
    pub fn send(&mut self, app_event: AppEvent) {
        // Ignore the result as the reciever cannot be dropped while this struct still has a
        // reference to it
        let _ = self.sender.send(Event::App(app_event).into());
    }

    /// Queue an app event to be sent to the event receiver (async version).
    ///
    /// This is useful for sending events from async contexts where you need proper error handling.
    pub async fn send_async(&self, app_event: AppEvent) -> color_eyre::Result<()> {
        self.sender
            .send(Event::App(app_event).into())
            .map_err(|_| color_eyre::eyre::eyre!("Failed to send event"))?;
        Ok(())
    }

    /// Queue a key event to be sent to the event receiver as a crossterm event.
    ///
    /// This is useful for programmatically sending key events that will be processed
    /// through the normal key event handling flow.
    pub fn send_key_event(&mut self, key_event: KeyEvent) {
        let crossterm_event = CrosstermEvent::Key(key_event);
        let _ = self.sender.send(Event::Crossterm(crossterm_event).into());
    }

    /// Check if there are any pending events in the queue.
    /// This is useful for testing to verify that events have been sent.
    #[cfg(test)]
    pub fn has_pending_events(&self) -> bool {
        !self.receiver.is_empty()
    }

    /// Try to receive an event without blocking.
    /// Returns None if no events are available.
    /// This is useful for testing to check what events have been sent.
    #[cfg(test)]
    pub fn try_recv(&mut self) -> Option<Arc<Event>> {
        self.receiver.try_recv().ok()
    }
}

/// A thread that handles reading crossterm events and emitting tick events on a regular schedule.
struct EventTask {
    /// Event sender channel.
    sender: mpsc::UnboundedSender<Arc<Event>>,
}

impl EventTask {
    /// Constructs a new instance of [`Event`].
    fn new(sender: mpsc::UnboundedSender<Arc<Event>>) -> Self {
        Self { sender }
    }

    /// Runs the event thread.
    ///
    /// This function emits tick events at a fixed rate and polls for crossterm events in between.
    async fn run(self) -> color_eyre::Result<()> {
        let tick_rate = Duration::from_secs_f64(1.0 / TICK_FPS);
        let mut reader = crossterm::event::EventStream::new();
        let mut tick = tokio::time::interval(tick_rate);
        loop {
            let tick_delay = tick.tick();
            let crossterm_event = reader.next().fuse();
            tokio::select! {
              _ = self.sender.closed() => {
                break;
              }
              _ = tick_delay => {
                self.send(Event::Tick.into());
              }
              Some(Ok(event)) = crossterm_event => {
                self.send(Event::Crossterm(event).into());
              }
            };
        }
        Ok(())
    }

    /// Sends an event to the receiver.
    fn send(&self, event: Arc<Event>) {
        // Ignores the result because shutting down the app drops the receiver, which causes the send
        // operation to fail. This is expected behavior and should not panic.
        let _ = self.sender.send(event);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ratatui::crossterm::event::{KeyCode, KeyEventKind, KeyEventState, KeyModifiers};

    #[test]
    fn test_send_key_event() {
        let mut event_handler = EventHandler::new_for_test();

        let key_event = KeyEvent {
            code: KeyCode::Char('q'),
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        // Send a key event
        event_handler.send_key_event(key_event);

        // Verify the event was sent as a crossterm event
        assert!(event_handler.has_pending_events());
        let event = event_handler.try_recv().unwrap();

        match *event {
            Event::Crossterm(CrosstermEvent::Key(received_key_event)) => {
                assert_eq!(received_key_event.code, KeyCode::Char('q'));
                assert_eq!(received_key_event.modifiers, KeyModifiers::empty());
                assert_eq!(received_key_event.kind, KeyEventKind::Press);
                assert_eq!(received_key_event.state, KeyEventState::empty());
            }
            _ => panic!("Expected crossterm key event, got: {event:?}"),
        }
    }
}
