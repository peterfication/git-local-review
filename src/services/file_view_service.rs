use std::{future::Future, pin::Pin, sync::Arc};

use crate::{
    database::Database,
    event::{AppEvent, EventHandler, ReviewId},
    models::FileView,
    services::{ServiceContext, ServiceHandler},
};

/// Service for handling file view operations
pub struct FileViewService;

impl ServiceHandler for FileViewService {
    fn handle_app_event<'a>(
        event: &'a AppEvent,
        context: ServiceContext<'a>,
    ) -> Pin<Box<dyn Future<Output = color_eyre::Result<()>> + Send + 'a>> {
        Box::pin(async move {
            match event {
                AppEvent::FileViewToggle {
                    review_id,
                    file_path,
                } => {
                    Self::handle_file_view_toggle(
                        context.database,
                        context.events,
                        review_id,
                        file_path,
                    )
                    .await?;
                }
                AppEvent::FileViewsLoad { review_id } => {
                    Self::handle_file_views_load(context.database, context.events, review_id)
                        .await?;
                }
                _ => {
                    // Event not handled by this service
                }
            }
            Ok(())
        })
    }
}

impl FileViewService {
    /// Toggle the view status of a file for a review
    async fn handle_file_view_toggle(
        database: &Database,
        events: &mut EventHandler,
        review_id: &ReviewId,
        file_path: &str,
    ) -> color_eyre::Result<()> {
        let pool = database.pool();

        // Check if file is currently viewed
        let is_viewed = match FileView::is_file_viewed(pool, review_id, file_path).await {
            Ok(viewed) => viewed,
            Err(error) => {
                events.send(AppEvent::FileViewToggleError {
                    review_id: Arc::from(review_id),
                    file_path: Arc::from(file_path),
                    error: Arc::from(format!("Failed to check file view status: {error}")),
                });
                return Ok(());
            }
        };

        let result = if is_viewed {
            // Mark as unviewed
            FileView::mark_as_unviewed(pool, review_id, file_path).await
        } else {
            // Mark as viewed
            FileView::mark_as_viewed(pool, review_id, file_path).await
        };

        match result {
            Ok(()) => {
                // Send success event and reload file views
                events.send(AppEvent::FileViewToggled {
                    review_id: Arc::from(review_id),
                    file_path: Arc::from(file_path),
                    is_viewed: !is_viewed,
                });
                events.send(AppEvent::FileViewsLoad {
                    review_id: Arc::from(review_id),
                });
            }
            Err(e) => {
                events.send(AppEvent::FileViewToggleError {
                    review_id: Arc::from(review_id),
                    file_path: Arc::from(file_path),
                    error: Arc::from(format!("Failed to toggle file view: {e}")),
                });
            }
        }

        Ok(())
    }

    /// Load the viewed files for a review
    async fn handle_file_views_load(
        database: &Database,
        events: &mut EventHandler,
        review_id: &ReviewId,
    ) -> color_eyre::Result<()> {
        let pool = database.pool();

        events.send(AppEvent::FileViewsLoading {
            review_id: Arc::from(review_id),
        });

        match FileView::get_viewed_files(pool, review_id).await {
            Ok(viewed_files) => {
                events.send(AppEvent::FileViewsLoaded {
                    review_id: Arc::from(review_id),
                    viewed_files: Arc::from(viewed_files),
                });
            }
            Err(e) => {
                events.send(AppEvent::FileViewsLoadError {
                    review_id: Arc::from(review_id),
                    error: Arc::from(format!("Failed to load file views: {e}")),
                });
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use sqlx::SqlitePool;

    use crate::{
        app::App,
        database::Database,
        event::{Event, EventHandler},
        models::Review,
    };

    async fn create_test_database() -> Database {
        let pool = SqlitePool::connect("sqlite::memory:").await.unwrap();
        sqlx::migrate!().run(&pool).await.unwrap();
        Database::from_pool(pool)
    }

    async fn create_test_review(database: &Database) -> Review {
        let review = Review::test_review(());
        review.save(database.pool()).await.unwrap();
        review
    }

    #[tokio::test]
    async fn test_handle_file_view_toggle_mark_as_viewed() {
        let database = create_test_database().await;
        let mut events = EventHandler::new_for_test();
        let review = create_test_review(&database).await;
        let file_path = "src/main.rs";

        // Ensure file is not viewed initially
        assert!(
            !FileView::is_file_viewed(database.pool(), &review.id, file_path)
                .await
                .unwrap()
        );

        // Toggle file view (should mark as viewed)
        let event = AppEvent::FileViewToggle {
            review_id: Arc::from(review.id.as_str()),
            file_path: Arc::from(file_path),
        };

        let app = App {
            running: true,
            events: EventHandler::new_for_test(),
            database,
            view_stack: vec![],
            repo_path: ".".to_string(),
        };
        FileViewService::handle_app_event(
            &event,
            ServiceContext {
                database: &app.database,
                repo_path: &app.repo_path,
                events: &mut events,
            },
        )
        .await
        .unwrap();

        // Verify file is now viewed
        assert!(
            FileView::is_file_viewed(app.database.pool(), &review.id, file_path)
                .await
                .unwrap()
        );

        // Check that success event was sent
        let sent_event = events.try_recv().unwrap();
        match &*sent_event {
            Event::App(AppEvent::FileViewToggled {
                review_id,
                file_path: sent_file_path,
                is_viewed,
            }) => {
                assert_eq!(review_id.as_ref(), review.id);
                assert_eq!(sent_file_path.as_ref(), file_path);
                assert!(*is_viewed);
            }
            _ => panic!("Expected FileViewToggled event, got: {sent_event:?}"),
        }

        // Check that file views reload event was sent
        let reload_event = events.try_recv().unwrap();
        match &*reload_event {
            Event::App(AppEvent::FileViewsLoad { review_id }) => {
                assert_eq!(review_id.as_ref(), review.id);
            }
            _ => panic!("Expected FileViewsLoad event, got: {reload_event:?}"),
        }
    }

    #[tokio::test]
    async fn test_handle_file_view_toggle_mark_as_unviewed() {
        let database = create_test_database().await;
        let mut events = EventHandler::new_for_test();
        let review = create_test_review(&database).await;
        let file_path = "src/main.rs";

        // First mark as viewed
        FileView::mark_as_viewed(database.pool(), &review.id, file_path)
            .await
            .unwrap();
        assert!(
            FileView::is_file_viewed(database.pool(), &review.id, file_path)
                .await
                .unwrap()
        );

        // Toggle file view (should mark as unviewed)
        let event = AppEvent::FileViewToggle {
            review_id: Arc::from(review.id.as_str()),
            file_path: Arc::from(file_path),
        };

        let app = App {
            running: true,
            events: EventHandler::new_for_test(),
            database,
            view_stack: vec![],
            repo_path: ".".to_string(),
        };
        FileViewService::handle_app_event(
            &event,
            ServiceContext {
                database: &app.database,
                repo_path: &app.repo_path,
                events: &mut events,
            },
        )
        .await
        .unwrap();

        // Verify file is now unviewed
        assert!(
            !FileView::is_file_viewed(app.database.pool(), &review.id, file_path)
                .await
                .unwrap()
        );

        // Check that success event was sent
        let sent_event = events.try_recv().unwrap();
        match &*sent_event {
            Event::App(AppEvent::FileViewToggled {
                review_id,
                file_path: sent_file_path,
                is_viewed,
            }) => {
                assert_eq!(review_id.as_ref(), review.id);
                assert_eq!(sent_file_path.as_ref(), file_path);
                assert!(!*is_viewed);
            }
            _ => panic!("Expected FileViewToggled event, got: {sent_event:?}"),
        }
    }

    #[tokio::test]
    async fn test_handle_file_views_load_success() {
        let database = create_test_database().await;
        let mut events = EventHandler::new_for_test();
        let review = create_test_review(&database).await;

        let file_paths = vec!["src/main.rs", "src/lib.rs"];
        for file_path in &file_paths {
            FileView::mark_as_viewed(database.pool(), &review.id, file_path)
                .await
                .unwrap();
        }

        // Load file views
        let event = AppEvent::FileViewsLoad {
            review_id: Arc::from(review.id.as_str()),
        };

        let app = App {
            running: true,
            events: EventHandler::new_for_test(),
            database,
            view_stack: vec![],
            repo_path: ".".to_string(),
        };
        FileViewService::handle_app_event(
            &event,
            ServiceContext {
                database: &app.database,
                repo_path: &app.repo_path,
                events: &mut events,
            },
        )
        .await
        .unwrap();

        // Check loading event was sent
        let loading_event = events.try_recv().unwrap();
        match &*loading_event {
            Event::App(AppEvent::FileViewsLoading { review_id }) => {
                assert_eq!(review_id.as_ref(), review.id);
            }
            _ => panic!("Expected FileViewsLoading event, got: {loading_event:?}"),
        }

        // Check loaded event was sent
        let loaded_event = events.try_recv().unwrap();
        match &*loaded_event {
            Event::App(AppEvent::FileViewsLoaded {
                review_id,
                viewed_files,
            }) => {
                assert_eq!(review_id.as_ref(), review.id);
                assert_eq!(viewed_files.len(), 2);
                assert!(viewed_files.contains(&"src/main.rs".to_string()));
                assert!(viewed_files.contains(&"src/lib.rs".to_string()));
            }
            _ => panic!("Expected FileViewsLoaded event, got: {loaded_event:?}"),
        }
    }

    #[tokio::test]
    async fn test_handle_file_views_load_empty() {
        let database = create_test_database().await;
        let mut events = EventHandler::new_for_test();
        let review = create_test_review(&database).await;

        // Load file views (none marked as viewed)
        let event = AppEvent::FileViewsLoad {
            review_id: Arc::from(review.id.as_str()),
        };

        let app = App {
            running: true,
            events: EventHandler::new_for_test(),
            database,
            view_stack: vec![],
            repo_path: ".".to_string(),
        };
        FileViewService::handle_app_event(
            &event,
            ServiceContext {
                database: &app.database,
                repo_path: &app.repo_path,
                events: &mut events,
            },
        )
        .await
        .unwrap();

        // Skip loading event
        events.try_recv().unwrap();

        // Check loaded event was sent with empty list
        let loaded_event = events.try_recv().unwrap();
        match &*loaded_event {
            Event::App(AppEvent::FileViewsLoaded {
                review_id,
                viewed_files,
            }) => {
                assert_eq!(review_id.as_ref(), review.id);
                assert_eq!(viewed_files.len(), 0);
            }
            _ => panic!("Expected FileViewsLoaded event, got: {loaded_event:?}"),
        }
    }

    #[tokio::test]
    async fn test_handle_unrelated_event() {
        let database = create_test_database().await;
        let mut events = EventHandler::new_for_test();

        // Send an unrelated event
        let event = AppEvent::Quit;

        let app = App {
            running: true,
            events: EventHandler::new_for_test(),
            database,
            view_stack: vec![],
            repo_path: ".".to_string(),
        };
        FileViewService::handle_app_event(
            &event,
            ServiceContext {
                database: &app.database,
                repo_path: &app.repo_path,
                events: &mut events,
            },
        )
        .await
        .unwrap();

        // No events should be sent
        assert!(!events.has_pending_events());
    }
}
