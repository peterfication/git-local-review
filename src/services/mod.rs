use std::future::Future;

use crate::database::Database;
use crate::event::{AppEvent, EventHandler};

pub mod git_service;
pub mod review_service;

pub use git_service::GitBranchesLoadingState;
pub use git_service::GitDiffLoadingState;
pub use git_service::GitService;
pub use review_service::ReviewCreateData;
pub use review_service::ReviewService;
pub use review_service::ReviewsLoadingState;

/// Trait for services that can handle app events
pub trait ServiceHandler {
    /// Handle an app event and potentially send new events through the event handler
    fn handle_app_event<'a>(
        event: &'a AppEvent,
        database: &'a Database,
        events: &'a mut EventHandler,
    ) -> std::pin::Pin<Box<dyn Future<Output = color_eyre::Result<()>> + Send + 'a>>;
}
