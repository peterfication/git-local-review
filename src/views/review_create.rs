use crate::{
    app::App,
    event::AppEvent,
    services::review_service::ReviewCreateData,
    views::{ViewHandler, ViewType},
};
use ratatui::{
    buffer::Buffer,
    crossterm::event::{KeyCode, KeyEvent},
    layout::{Constraint, Direction, Layout, Rect},
    style::{Color, Style},
    widgets::{Block, BorderType, Clear, Paragraph, Widget},
};

#[derive(Default)]
pub struct ReviewCreateView {
    pub title_input: String,
}

impl ViewHandler for ReviewCreateView {
    fn view_type(&self) -> ViewType {
        ViewType::ReviewCreate
    }

    fn handle_key_events(&mut self, app: &mut App, key_event: KeyEvent) -> color_eyre::Result<()> {
        match key_event.code {
            KeyCode::Esc => {
                self.title_input.clear();
                app.events.send(AppEvent::ReviewCreateClose);
            }
            KeyCode::Enter => {
                app.events
                    .send(AppEvent::ReviewCreateSubmit(ReviewCreateData {
                        title: self.title_input.clone(),
                    }));
                self.title_input.clear();
            }
            KeyCode::Char(char) => {
                self.title_input.push(char);
            }
            KeyCode::Backspace => {
                self.title_input.pop();
            }
            _ => {}
        }
        Ok(())
    }

    fn render(&self, _app: &App, area: Rect, buf: &mut Buffer) {
        let popup_area = centered_rect(60, 40, area);

        Clear.render(popup_area, buf);

        let block = Block::bordered()
            .title("Create New Review")
            .border_type(BorderType::Rounded)
            .style(Style::default().bg(Color::Black));

        let inner = block.inner(popup_area);
        block.render(popup_area, buf);

        let chunks = Layout::default()
            .direction(Direction::Vertical)
            .constraints([
                Constraint::Length(1),
                Constraint::Length(3),
                Constraint::Length(1),
            ])
            .split(inner);

        let title_label = Paragraph::new("Title:");
        title_label.render(chunks[0], buf);

        let title_input = Paragraph::new(self.title_input.as_str())
            .block(Block::bordered())
            .style(Style::default().fg(Color::Yellow));
        title_input.render(chunks[1], buf);

        let help = Paragraph::new("Press Enter to create, Esc to cancel")
            .style(Style::default().fg(Color::Gray));
        help.render(chunks[2], buf);
    }

    #[cfg(test)]
    fn debug_state(&self) -> String {
        format!("ReviewCreateView(title_input: \"{}\")", self.title_input)
    }

    #[cfg(test)]
    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
        self
    }
}

fn centered_rect(percent_x: u16, percent_y: u16, r: Rect) -> Rect {
    let popup_layout = Layout::default()
        .direction(Direction::Vertical)
        .constraints([
            Constraint::Percentage((100 - percent_y) / 2),
            Constraint::Percentage(percent_y),
            Constraint::Percentage((100 - percent_y) / 2),
        ])
        .split(r);

    Layout::default()
        .direction(Direction::Horizontal)
        .constraints([
            Constraint::Percentage((100 - percent_x) / 2),
            Constraint::Percentage(percent_x),
            Constraint::Percentage((100 - percent_x) / 2),
        ])
        .split(popup_layout[1])[1]
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::database::Database;
    use crate::event::{AppEvent, Event};
    use crate::models::review::Review;
    use crate::services::ReviewsLoadingState;
    use crate::test_utils::render_app_to_terminal_backend;
    use insta::assert_snapshot;
    use sqlx::SqlitePool;

    use ratatui::crossterm::event::{KeyCode, KeyEvent, KeyEventKind, KeyEventState, KeyModifiers};

    async fn create_test_app() -> App {
        let pool = SqlitePool::connect("sqlite::memory:").await.unwrap();
        Review::create_table(&pool).await.unwrap();

        let database = Database::from_pool(pool);
        let reviews = vec![];

        App {
            running: true,
            events: crate::event::EventHandler::new_for_test(),
            database,
            reviews,
            reviews_loading_state: ReviewsLoadingState::Loaded,
            view_stack: vec![],
        }
    }

    #[test]
    fn test_review_create_view_default() {
        let view = ReviewCreateView::default();
        assert_eq!(view.title_input, "");
    }

    #[tokio::test]
    async fn test_review_create_view_handle_char_input() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView::default();

        let key_event = KeyEvent {
            code: KeyCode::Char('a'),
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        assert_eq!(view.title_input, "a");
    }

    #[tokio::test]
    async fn test_review_create_view_handle_multiple_chars() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView::default();

        let chars = ['H', 'e', 'l', 'l', 'o'];
        for c in chars {
            let key_event = KeyEvent {
                code: KeyCode::Char(c),
                modifiers: KeyModifiers::empty(),
                kind: KeyEventKind::Press,
                state: KeyEventState::empty(),
            };
            view.handle_key_events(&mut app, key_event).unwrap();
        }

        assert_eq!(view.title_input, "Hello");
    }

    #[tokio::test]
    async fn test_review_create_view_handle_backspace() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView {
            title_input: "Hello".to_string(),
        };

        let key_event = KeyEvent {
            code: KeyCode::Backspace,
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        assert_eq!(view.title_input, "Hell");
    }

    #[tokio::test]
    async fn test_review_create_view_handle_backspace_empty() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView::default();

        let key_event = KeyEvent {
            code: KeyCode::Backspace,
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        assert_eq!(view.title_input, "");
    }

    #[tokio::test]
    async fn test_review_create_view_handle_esc() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView {
            title_input: "Some input".to_string(),
        };
        assert!(!app.events.has_pending_events());

        let key_event = KeyEvent {
            code: KeyCode::Esc,
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        // Title input should be cleared
        assert_eq!(view.title_input, "");

        // Verify that a ReviewCreateClose event was sent
        assert!(app.events.has_pending_events());
        let event = app.events.try_recv().unwrap();
        assert!(matches!(event, Event::App(AppEvent::ReviewCreateClose)));
    }

    #[tokio::test]
    async fn test_review_create_view_handle_enter() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView {
            title_input: "Test Review".to_string(),
        };
        assert!(!app.events.has_pending_events());

        let key_event = KeyEvent {
            code: KeyCode::Enter,
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        // Title input should be cleared after submit
        assert_eq!(view.title_input, "");

        // Verify that a ReviewCreateSubmit event was sent with the correct title
        assert!(app.events.has_pending_events());
        let event = app.events.try_recv().unwrap();
        if let Event::App(AppEvent::ReviewCreateSubmit(data)) = event {
            assert_eq!(data.title, "Test Review");
        } else {
            panic!("Expected ReviewCreateSubmit event");
        }
    }

    #[tokio::test]
    async fn test_review_create_view_handle_enter_empty() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView::default();

        let key_event = KeyEvent {
            code: KeyCode::Enter,
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        // Should still work with empty input
        assert_eq!(view.title_input, "");
    }

    #[tokio::test]
    async fn test_review_create_view_handle_unknown_key() {
        let mut app = create_test_app().await;
        let mut view = ReviewCreateView::default();
        let initial_input = "Test".to_string();
        view.title_input = initial_input.clone();

        let key_event = KeyEvent {
            code: KeyCode::F(1),
            modifiers: KeyModifiers::empty(),
            kind: KeyEventKind::Press,
            state: KeyEventState::empty(),
        };

        view.handle_key_events(&mut app, key_event).unwrap();

        // Unknown keys should not change input
        assert_eq!(view.title_input, initial_input);
    }

    #[tokio::test]
    async fn test_review_create_view_render_default() {
        let view = ReviewCreateView::default();
        let app = App {
            view_stack: vec![Box::new(view)],
            ..create_test_app().await
        };

        assert_snapshot!(render_app_to_terminal_backend(app))
    }

    #[tokio::test]
    async fn test_review_create_view_render_with_title() {
        let view = ReviewCreateView {
            title_input: "My New Review".to_string(),
        };
        let app = App {
            view_stack: vec![Box::new(view)],
            ..create_test_app().await
        };

        assert_snapshot!(render_app_to_terminal_backend(app))
    }
}
