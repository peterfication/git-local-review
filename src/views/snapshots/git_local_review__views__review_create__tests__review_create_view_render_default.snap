---
source: src/views/review_create.rs
expression: render_app_to_terminal_backend(app)
---
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                ╭Create New Review─────────────────────────────────────────────────────────────────────────────╮                                "
"                                │Title:                                                                                        │                                "
"                                │┌────────────────────────────────────────────────────────────────────────────────────────────┐│                                "
"                                ││                                                                                            ││                                "
"                                │└────────────────────────────────────────────────────────────────────────────────────────────┘│                                "
"                                │Press Enter to create, Esc to cancel                                                          │                                "
"                                │                                                                                              │                                "
"                                │                                                                                              │                                "
"                                │                                                                                              │                                "
"                                │                                                                                              │                                "
"                                │                                                                                              │                                "
"                                │                                                                                              │                                "
"                                │                                                                                              │                                "
"                                │                                                                                              │                                "
"                                │                                                                                              │                                "
"                                ╰──────────────────────────────────────────────────────────────────────────────────────────────╯                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
"                                                                                                                                                                "
